README
What is it?
===========

Xfce panel plugin which allows to put the maximized window title and windows buttons on the panel.

It's main features are:

- Show the title and buttons of the maximized window on the panel.
- Allow window actions on buttons and title clicks (activate, (un)maximize, close).
- Allow window action menu on left button click.
- Title formatting options.
- xfwm4 theming support for buttons.

FAQ:
Q: How do I hide the window decorations on maximized windows on xfce ?
A: You can use a patch for xfwm4 available here : https://github.com/cedl38/xfwm4-titleless. Alternatively, you can try to use an external program like maximus (performance should be lower).
There are also some xfwm4 themes with very thin border like this one :
http://xfce-look.org/content/show.php?content=139915.

This code is derived from original 'Window Applets' from Andrej Belcijan.
See http://gnome-look.org/content/show.php?content=103732 for infos.
 
Installation
============

The file 'INSTALL' contains generic installation instructions. For more
detailed information, visit the plugin website at
http://goodies.xfce.org/projects/panel-plugins/xfce4-windowck-plugin


How to report bugs?
===================

Bugs should be reported to the GitHub project
(https://github.com/cedl38/xfce4-windowck-plugin). You will need to create an account for yourself.


NEWS
0.5.0
	Fix #72 Port to GTK 3
0.4.8
	Bumped compatibility for Debian 10 / Xubuntu 20.04
0.4.7
	Fix crash of title plugin when show on desktop (by mskalski)
	Fix #30 Fixed memory leak (by invidian)
0.4.6
    Fix #51: Build warnings
    Fix #56: add some checks while processing changes of unfocused window title
    Fix #64 Added Russian translation
    Fix #71 Fix crash on closing last window
0.4.5
    Fix #28 - Updated deb packages dependency for ubuntu 15.10+
0.4.4
    Fix #22 support of *.svg based unity themes
0.4.3
    Fix #21 disable activate mouse action on desktop
    properly hide title on desktop #21
0.4.2
    ReAdd missing themerc from windowck themes
0.4.1
    Add missing dependency in PKGBUILD
    Update debian packaging
    Remove *.xpm files from source code
0.4.0
    Fix Icon on right behaviour reg. #da3b6c6
    Drop xpm support
    Improve windowck themes
    Added Czech translation 100%
    fix #16 Remove background on buttons
    fix #14 correctly update title when switching workspace
    fix #5 Let windowck plugin skip whiskermenu
0.3.1
    Allow 'Log Out on close desktop' #12
    Add multi-arch support to Debian packaging
0.3.0
    Prepare release 0.3.0
    Updates french translations (100%) and italian (45%)
    Allow title text to be displayed on two lines
    Allow resizing the window icon
    Fix DEFAULT_SYNC_WM_FONT case formating
    Fix #7 Window title widget vertical align
0.2.5
    Add option to sync font with window manager for title plugin
    Change show_window_menu option behaviour
    Fix gcc warnings
    Fix inactive title color for dark themes
0.2.4
    Update plugin when the gtk theme changed
    Fix fg color issue with some themes
    Fix autogen warnings
0.2.3
    Fix a wrong title display when desktop is shown.
0.2.2
    Fix #2 plugin initialization issue
0.2.1
    Disable actions on desktop window.
    Fix wck_utils to properly disconnect signals.
    Fix missing PKGBUILD.in.
    Fix gcc warnings.
    Fix xfwm4 theme loading.
0.2.0
    Provide compatibility with unity button themes.
    Add option to sync the theme with the window manager.
    Add theme selection and button layout dialog options.
    The menu button can show an arrow instead of the window icon.
    Add option to show/hide the menu button.
    Change dialogs windows appearance.
    Fix reloadplugins.sh

0.1.3
    Backport 0.2.2 updates.

0.1.1
    Add tooltip on size_mode combobox and correct a misspelling.
    Fix maximize button action.
    Fix window tracking issue when desktop displayed and others special cases.
    Fix theme loading.
    Update Italian and French translations.

0.1.0
    Improve controlled window state tracking.
    Add wckbuttons plugin.
    Add window actions on buttons and title clicks (activate, (un)maximize, close).
    Add window action menu on left button click.
    Set inactive title color to fit better to themes.
    Add optional custom title font.
    Add xfwm4 theme support for buttons.
    Add plugins icons and a few themes.
    Add window icon to windowck plugin.
    Fix title overflow for expand option (use xfce_panel_plugin_set_shrink).

0.0.2
    Add shrink, fixed and expand title options.
    Add title padding and alignment options.

0.0.1
    Initial release.


panel-plugin/buttons/theme.c
/*  $Id$
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 *  Copyright (C) 2013 Cedric Leporcq  <cedl38@gmail.com>
 *
 *  This stuff is mostly derived from xfwm4 sources (setting.c)
 *
 */

#include <libxfce4util/libxfce4util.h>
#include <common/ui_style.h>
#include <common/mypixmap.h>

#include "theme.h"

#define XPM_COLOR_SYMBOL_SIZE 22
#define THEMERC "themerc"

gchar *
getSystemThemeDir (void)
{
    return g_build_filename (DATADIR, "themes", DEFAULT_THEME, "xfwm4", NULL);
}

gchar *
getThemeDir (const gchar * theme, const gchar * file)
{
    static const char *themedir[] = {
      "windowck",
      "xfwm4"
    };

    if (!theme)
    {
        return g_build_filename (DATADIR, "themes", DEFAULT_THEME, "xfwm4",
                                 NULL);
    }
    else if (g_path_is_absolute (theme))
    {
        if (g_file_test (theme, G_FILE_TEST_IS_DIR))
        {
            return g_strdup (theme);
        }
        else
        {
            return getSystemThemeDir ();
        }
    }
    else
    {
        gchar *test_file, *path;
        gint i;

        for (i=0; i<2; i++) {

            path = g_build_filename (theme, themedir[i], file, NULL);

            xfce_resource_push_path (XFCE_RESOURCE_THEMES,
                                     DATADIR G_DIR_SEPARATOR_S "themes");
            test_file = xfce_resource_lookup (XFCE_RESOURCE_THEMES, path);
            xfce_resource_pop_path (XFCE_RESOURCE_THEMES);

            g_free (path);

            if (test_file)
            {
                path = g_path_get_dirname (test_file);
                g_free (test_file);
                return path;
            }
        }
    }

    /* Pfew, really can't find that theme nowhere! */
    return getSystemThemeDir ();
}

static gboolean
setGValue (const gchar * lvalue, const GValue *rvalue, Settings *rc)
{
    gint i;

    TRACE ("entering setValue");

    g_return_val_if_fail (lvalue != NULL, FALSE);
    g_return_val_if_fail (rvalue != NULL, FALSE);

    for (i = 0; rc[i].option; i++)
    {
        if (!g_ascii_strcasecmp (lvalue, rc[i].option))
        {
            if (rvalue)
            {
                if (rc[i].value)
                {
                    g_value_unset (rc[i].value);
                    g_value_init (rc[i].value, G_VALUE_TYPE(rvalue));
                }
                else
                {
                    rc[i].value = g_new0(GValue, 1);
                    g_value_init (rc[i].value, G_VALUE_TYPE(rvalue));
                }

                g_value_copy (rvalue, rc[i].value);
                return TRUE;
            }
        }
    }
    return FALSE;
}

static gboolean
setStringValue (const gchar * lvalue, const gchar *value, Settings *rc)
{
    GValue tmp_val = {0, };
    g_value_init(&tmp_val, G_TYPE_STRING);
    g_value_set_static_string(&tmp_val, value);
    return setGValue (lvalue, &tmp_val, rc);
}

/* load the theme according to xfwm4 theme format */
void loadTheme (WBPlugin *wb)
{
    Settings rc[] = {
        /* Do not change the order of the following parameters */
        {"active_text_color", NULL, G_TYPE_STRING, FALSE},
        {"inactive_text_color", NULL, G_TYPE_STRING, FALSE},
        {"active_text_shadow_color", NULL, G_TYPE_STRING, FALSE},
        {"inactive_text_shadow_color", NULL, G_TYPE_STRING, FALSE},
        {"active_border_color", NULL, G_TYPE_STRING, FALSE},
        {"inactive_border_color", NULL, G_TYPE_STRING, FALSE},
        {"active_color_1", NULL, G_TYPE_STRING, FALSE},
        {"active_hilight_1", NULL, G_TYPE_STRING, FALSE},
        {"active_shadow_1", NULL, G_TYPE_STRING, FALSE},
        {"active_mid_1", NULL, G_TYPE_STRING, FALSE},
        {"active_color_2", NULL, G_TYPE_STRING, FALSE},
        {"active_hilight_2", NULL, G_TYPE_STRING, FALSE},
        {"active_shadow_2", NULL, G_TYPE_STRING, FALSE},
        {"active_mid_2", NULL, G_TYPE_STRING, FALSE},
        {"inactive_color_1", NULL, G_TYPE_STRING, FALSE},
        {"inactive_hilight_1", NULL, G_TYPE_STRING, FALSE},
        {"inactive_shadow_1", NULL, G_TYPE_STRING, FALSE},
        {"inactive_mid_1", NULL, G_TYPE_STRING, FALSE},
        {"inactive_color_2", NULL, G_TYPE_STRING, FALSE},
        {"inactive_hilight_2", NULL, G_TYPE_STRING, FALSE},
        {"inactive_shadow_2", NULL, G_TYPE_STRING, FALSE},
        {"inactive_mid_2", NULL, G_TYPE_STRING, FALSE}
    };

    static const char *ui_part[] = {
        "text",
        "mix_bg_text",
        "dark",
        "dark",
        "fg",
        "fg",
        "bg",
        "light",
        "dark",
        "mid",
        "bg",
        "light",
        "dark",
        "mid",
        "bg",
        "light",
        "dark",
        "mid",
        "bg",
        "light",
        "dark",
        "mid",
        NULL
    };

    static const char *ui_state[] = {
        "active",
        "normal",
        "active",
        "insensitive",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        "normal",
        NULL
    };

    /* use xfwm4 buttons naming */
    static const char *button_names[] = {
      //~ "menu",
      //~ "stick",
      //~ "shade",
      "hide",
      "maximize-toggled",
      "maximize",
      "close"
    };

    static const char *button_state_names[] = {
      "inactive",
      "active",
      "prelight",
      "pressed"
    };

    gint i,j;
    gchar imagename[30];
    gchar *theme;
    xfwmColorSymbol colsym[ XPM_COLOR_SYMBOL_SIZE + 1 ];

    for (i = 0; ui_part[i] && ui_state[i]; i++)
    {
        gchar *color;

        color = getUIStyle  (GTK_WIDGET(wb->plugin), ui_part[i], ui_state[i]);
        setStringValue (rc[i].option, color, rc);
        g_free (color);
    }

    for (i = 0; i < XPM_COLOR_SYMBOL_SIZE; i++)
    {
        colsym[i].name = rc[i].option;
        colsym[i].value = g_value_get_string(rc[i].value);
    }
    colsym[XPM_COLOR_SYMBOL_SIZE].name = NULL;
    colsym[XPM_COLOR_SYMBOL_SIZE].value = NULL;

    /* get theme filename */
    theme = getThemeDir (wb->prefs->theme, THEMERC);

    for (i = 0; i < IMAGES_BUTTONS; i++)
    {
        for (j = 0; j < IMAGES_STATES; j++)
        {
            g_snprintf(imagename, sizeof (imagename), "%s-%s", button_names[i], button_state_names[j]);
            wb->pixbufs[i][j] = xfwmPixbufLoad (theme, imagename, colsym);
        }
    }

    /* try to replace missing images */

    for (i = 0; i < BUTTONS; i++)
    {
        if (G_LIKELY(wb->pixbufs[i][IMAGE_FOCUSED]))
            wb->button[i]->visible = TRUE;
        else
            wb->button[i]->visible = FALSE;
    }

    for (i = 0; i < IMAGES_STATES; i++)
    {
        if (!wb->pixbufs[IMAGE_UNMAXIMIZE][i])
            wb->pixbufs[IMAGE_UNMAXIMIZE][i] = wb->pixbufs[IMAGE_MAXIMIZE][i];
    }
    for (i = 0; i < IMAGES_BUTTONS; i++)
    {
        if (!wb->pixbufs[i][IMAGE_UNFOCUSED] || !wb->pixbufs[i][IMAGE_PRESSED])
            wb->pixbufs[i][IMAGE_UNFOCUSED] = wb->pixbufs[i][IMAGE_FOCUSED];

        if (!wb->pixbufs[i][IMAGE_PRELIGHT])
            wb->pixbufs[i][IMAGE_PRELIGHT] = wb->pixbufs[i][IMAGE_PRESSED];
    }
}
